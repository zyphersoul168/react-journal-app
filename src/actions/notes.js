import Swal from "sweetalert2";

import { addDoc, deleteDoc, setDoc, collection, doc } from "firebase/firestore";
import { db } from "../firebase/config";
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";
import { fileUpload } from "../helpers/fileUpload";

export const startNewNote = () => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;

    const newNote = {
      title: "",
      body: "",
      date: new Date().getTime(),
    };

    const userRef = collection(db, `${uid}/journal/notes`);

    try {
      // add note to database
      const noteRef = await addDoc(userRef, newNote);

      dispatch(activeNote(noteRef.id, newNote));
      dispatch(addNewNote(noteRef.id, newNote));
    } catch (error) {
      console.log(error);
    }
  };
};

export const activeNote = (id, note) => ({
  type: types.notesActive,
  payload: {
    id,
    ...note,
  },
});

export const addNewNote = (id, note) => ({
  type: types.notesAddNew,
  payload: {
    id,
    ...note,
  },
});

export const startLoadingNotes = (uid) => {
  return async (dispatch) => {
    const notes = await loadNotes(uid);
    dispatch(setNotes(notes));
  };
};

export const setNotes = (notes) => ({
  type: types.notesLoad,
  payload: notes,
});

export const startSaveNote = (note) => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;

    if (!note.url) {
      delete note.url;
    }

    const noteToFirestore = { ...note };
    delete noteToFirestore.id;

    await setDoc(doc(db, `${uid}/journal/notes`, note.id), noteToFirestore);

    dispatch(refreshNote(note.id, noteToFirestore));
    Swal.fire("Saved!", note.title, "success");
  };
};

export const refreshNote = (id, note) => ({
  type: types.notesUpdate,
  payload: {
    id,
    note: {
      id,
      ...note,
    },
  },
});

export const startUploadingImage = (file) => {
  return async (dispatch, getState) => {
    const { active: activeNote } = getState().notes;

    Swal.fire({
      title: "Uploading...",
      text: "Wait :)",
      allowOutsideClick: false,
      didOpen: () => {
        Swal.showLoading();
      },
    });

    const fileUrl = await fileUpload(file);
    activeNote.url = fileUrl;

    dispatch(startSaveNote(activeNote));

    Swal.close();
  };
};

export const startDeletingNote = (noteId) => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;
    await deleteDoc(doc(db, `${uid}/journal/notes`, noteId));

    dispatch(deleteNote(noteId));

    Swal.fire("Deleted!", "Note deleted", "success");
  };
};

export const deleteNote = (noteId) => ({
  type: types.notesDelete,
  payload: noteId,
});

export const cleanNotes = () => ({
  type: types.notesLogoutClean,
});
