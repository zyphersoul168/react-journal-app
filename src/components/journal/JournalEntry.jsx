import React from 'react';

export const JournalEntry = () => {
  return (
    <div className="journal__entry pointer">
      <div
        className="journal__entry-picture"
        style={{
          backgroundSize: 'cover',
          backgroundImage:
            'url(https://pm1.narvii.com/7673/bc18bebccadba969afa1b0778c53bbdfa93bda26r1-540-553v2_hq.jpg)',
        }}></div>
      <div className="journal__entry-body">
        <p className="journal__entry-title">Un nuevo dÃ­a</p>
        <p className="journal__entry-content">
          Cillum tempor est esse incididunt ex velit est quis pariatur
          consectetur.
        </p>
      </div>
      <div className="journal__entry-date-box">
        <span>Monday</span>
        <h4>5</h4>
      </div>
    </div>
  );
};
