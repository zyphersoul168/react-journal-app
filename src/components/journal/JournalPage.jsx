import React from 'react'

export const JournalPage = () => {
  return (
    <div>
      <h1>JournalPage</h1>
    </div>
  )
}
