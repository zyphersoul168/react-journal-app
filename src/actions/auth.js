import {
  getAuth,
  signInWithPopup,
  createUserWithEmailAndPassword,
  updateProfile,
  signInWithEmailAndPassword,
  signOut,
} from 'firebase/auth';
import { googleAuthProvider } from '../firebase/config';
import { types } from '../types/types';
import { finishLoading, startLoading } from './ui';
import Swal from 'sweetalert2';
import { errors } from '../types/errors';
import { cleanNotes } from './notes';

const auth = getAuth();

export const startLoginEmailPassword = (email, password) => {
  return async (dispatch) => {
    dispatch(startLoading());
    try {
      const { user } = await signInWithEmailAndPassword(auth, email, password);
      dispatch(login(user.uid, user.displayName));
      dispatch(finishLoading());
    } catch (error) {
      dispatch(finishLoading());
      error.message === errors.userNotFound &&
        Swal.fire('Error', 'User not found', 'error');
      error.message === errors.wrongPassword &&
        Swal.fire('Error', 'Wrong Password', 'error');
    }
  };
};

export const startRegisterWithEmailPasswordName = (email, password, name) => {
  return (dispatch) => {
    createUserWithEmailAndPassword(auth, email, password)
      .then(async ({ user }) => {
        await updateProfile(user, { displayName: name });
        dispatch(login(user.uid, user.displayName));
      })
      .catch((error) => {
        error.message === errors.existingAccount &&
          Swal.fire('Error', 'There is already an account with this email', 'error');
        console.log(error.message);
      });
  };
};

export const startGoogleLogin = () => {
  return (dispatch) => {
    signInWithPopup(auth, googleAuthProvider).then(({ user }) => {
      dispatch(login(user.uid, user.displayName));
    });
  };
};

export const login = (uid, displayName) => ({
  type: types.login,
  payload: {
    uid,
    displayName,
  },
});

export const startLogout = () => {
  return async (dispatch) => {
    await signOut(auth);
    dispatch(logout());
    dispatch(cleanNotes());
  };
};

export const logout = () => ({
  type: types.logout,
});
