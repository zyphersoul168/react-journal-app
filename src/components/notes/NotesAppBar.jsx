import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { startDeletingNote, startSaveNote, startUploadingImage } from "../../actions/notes";

export const NotesAppBar = () => {
  const dispatch = useDispatch();
  const { active: activeNote } = useSelector((state) => state.notes);

  const handleSave = () => {
    dispatch(startSaveNote(activeNote));
    console.log("note updated");
  };

  const handlePictureUpload = () => {
    document.querySelector("#fileSelector").click();
  };

  const handleFileChange = (event) => {
    const file = event.target.files[0];

    if (file) dispatch(startUploadingImage(file));
  };

  const handleDelete = () => {
    dispatch(startDeletingNote(activeNote.id));
  }

  return (
    <div className="notes__appbar">
      <span>Septiembre 6, 2021</span>
      <input
        id="fileSelector"
        name="file"
        type="file"
        style={{ display: "none" }}
        onChange={handleFileChange}
      />
      <div>
        <button className="btn notes__appbar-btn" onClick={handlePictureUpload}>
          Picture
        </button>
        <button className="btn notes__appbar-btn" onClick={handleSave}>
          Save
        </button>
        <button className="btn btn-danger" onClick={handleDelete}>Delete</button>
      </div>
    </div>
  );
};
