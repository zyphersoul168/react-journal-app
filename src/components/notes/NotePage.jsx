import React, { useEffect, useRef } from "react";
import { useSelector, useDispatch } from "react-redux";
import { NotesAppBar } from "./NotesAppBar";
import { useForm } from "../../hooks/useForm";
import { activeNote } from "../../actions/notes";

export const NotePage = () => {
  const dispatch = useDispatch();
  const { active: note } = useSelector((state) => state.notes);
  const [formValues, handleInputChange, reset] = useForm(note);
  const { body, title } = formValues;
  const activeId = useRef(note.id);

  useEffect(() => {
    if(note.id !== activeId.current) {
      reset(note);
      activeId.current = note.id
    }
  }, [note, reset]);

  useEffect(() => {
    dispatch(activeNote(formValues.id, {...formValues}))
  }, [formValues, dispatch]);

  return (
    <div className="notes__main-content">
      <NotesAppBar />
      <div className="notes__content">
        <input
          type="text"
          className="notes__title-input"
          placeholder="An awesome history?"
          autoComplete="off"
          name="title"
          value={title}
          onChange={handleInputChange}
        />
        <textarea
          placeholder="Tell it here!"
          className="notes__textarea"
          name="body"
          value={body}
          onChange={handleInputChange}
        ></textarea>
        {note.url && (
          <div className="notes__img">
            <img
              src={note.url}
              alt="imagen"
            />
          </div>
        )}
      </div>
    </div>
  );
};
