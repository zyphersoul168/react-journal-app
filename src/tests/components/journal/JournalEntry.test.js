import React from "react";
import { mount } from "enzyme";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import "@testing-library/jest-dom";
import { JournalEntry } from "../../../components/journal/JournalEntry";
import { activeNote } from "../../../actions/notes";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initialState = {};
let store = mockStore(initialState);
store.dispatch = jest.fn();

const note = {
  id: "abc",
  date: 0,
  title: "Hola",
  body: "Mundo",
  url: "https://piolamano.com/photo.png",
};

const wrapper = mount(
  <Provider store={store}>
    <JournalEntry {...note} />
  </Provider>
);

describe("Tests on JournalEntry component", () => {
  test("should match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should dispatch activeNote action", () => {
    wrapper.find(".journal__entry").simulate("click");

    expect(store.dispatch).toHaveBeenCalledWith(
      activeNote(note.id, { ...note })
    );
  });
});
