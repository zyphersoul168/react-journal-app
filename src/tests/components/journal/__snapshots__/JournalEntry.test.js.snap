// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on JournalEntry component should match snapshot 1`] = `
<div
  className="journal__entry pointer"
  onClick={[Function]}
>
  <div
    className="journal__entry-picture"
    style={
      Object {
        "backgroundImage": "url(https://piolamano.com/photo.png)",
        "backgroundSize": "cover",
      }
    }
  />
  <div
    className="journal__entry-body"
  >
    <p
      className="journal__entry-title"
    >
      Hola
    </p>
    <p
      className="journal__entry-content"
    >
      Mundo
    </p>
  </div>
  <div
    className="journal__entry-date-box"
  >
    <span>
      Wednesday
    </span>
    <h4>
      31st
    </h4>
  </div>
</div>
`;
